graph.go

```diff
@@ -11,197 +11,162 @@ type Node interface {
 	ID() int
 }
 
-// Allows edges to do something more interesting that just be a group of nodes. While the methods
-// are called Head and Tail, they are not considered directed unless the given interface specifies
-// otherwise.
-type Edge interface {
+// An unordered group of two nodes
+type UndirectedEdge interface {
```

Comment:
I don't think that we need to define this. When you have a node, irrespective of whether you have a directed graph or an undirected graph there are outgoing edges from that node. The difference then is that an undirected graph returns all incident edges, wherase a directed graph only returns edges that face out. To handle the semantics of this, undirected graphs should always return an edge{u,v} for a call to Out(u) and always return edge{v,u} for a call to In(u).

This leaves the problem of how you distinguish between directed and undirected graphs - my api-explore branch shows how this can work; define EdgeBetween and EdgeFromTo methods. Now, we have that graph.Graph interface with the In and Out methods that you introduce here.